### PR TITLE
Allow signed root regeneration when quiescent

### PR DESCRIPTION
**Pull request by daviddrysdale**
- First commit includes the five line change that actually does something.
- Second commit is the two dozen places that need to be changed to plumb in a new per-Tree option
- Third, fourth and sixth commits are re-workings of existing code that were requested in previous code review (#579).

### CODE REVIEW

File: log/sequencer.go [hunk 1]
```diff
@@ -204,12 +207,17 @@ func (s Sequencer) SequenceBatch(ctx context.Context, logID int64, limit int) (i
...
 	}
 
-	// There might be no work to be done. But we possibly still need to create an STH if the
+	// There might be no work to be done. But we possibly still need to create an signed root if the
 	// current one is too old. If there's work to be done then we'll be creating a root anyway.
 	if len(leaves) == 0 {
-		// We have nothing to integrate into the tree
-		glog.V(1).Infof("No leaves sequenced in this signing operation.")
-		return 0, tx.Commit()
+		nowNanos := s.timeSource.Now().UnixNano()
```
Comment (AlCutter) [inline log/sequencer.go]: I think the signer framework uses a call into storage which, in some implementations at least, knows whether there's anything in the pending queue, so you may not even get in here if it knows there's no work to be done...

File: cmd/createtree/main.go [hunk 2]
```diff
@@ -125,14 +128,15 @@ func newRequest(opts *createOpts) (*trillian.CreateTreeRequest, error) {
...
-		PrivateKey:         pk,
+		TreeState:             trillian.TreeState(ts),
+		TreeType:              trillian.TreeType(tt),
+		HashStrategy:          trillian.HashStrategy(hs),
+		HashAlgorithm:         sigpb.DigitallySigned_HashAlgorithm(ha),
+		SignatureAlgorithm:    sigpb.DigitallySigned_SignatureAlgorithm(sa),
+		DisplayName:           opts.displayName,
+		Description:           opts.description,
+		PrivateKey:            pk,
+		MaxRootDurationMillis: opts.maxRootDuration.Nanoseconds() / int64(time.Millisecond),
```
Comment (AlCutter) [inline cmd/createtree/main.go]: `MaxRootIntervalMillis`?

Comment (daviddrysdale) [inline cmd/createtree/main.go]: This has already been renamed once so I'd prefer not to do so again.

File: log/sequencer.go (hunk 1 above)
Comment (daviddrysdale) [inline log/sequencer.go]: The `GetActiveLogIDsWithPendingWork` entrypoint doesn't look to be used (and doesn't have a non-trivial implementation anywhere), so I've added a commit to just remove it.

File: cmd/createtree/main.go (hunk 2 above)
Comment (AlCutter) [inline cmd/createtree/main.go]: What was the rationale?
`...Duration...` feels a little confusing to me, "is it how long the root is valid for, or ...", whereas `...Interval...` seems more likely to be understood as the maximum interval between roots - the help string for the flag starts "Interval after which...", same with the field documentation in the proto file...

File: log/sequencer.go (hunk 1 above)
Comment (AlCutter) [inline log/sequencer.go]: Nice one

File: cmd/createtree/main.go (hunk 2 above)
Comment (daviddrysdale) [inline cmd/createtree/main.go]: Original suggestion at https://github.com/google/trillian/pull/579#discussion_r114785561.

At this point I couldn't care less what the option is called, but if I have to change its name again I'm dropping the middle 3 commits...

**Comment by daviddrysdale** [comment]
Added an extra commit to convert `Tree` timing fields in `trillian.proto` to use `google.protobuf` time types.

File: integration/admin/admin_integration_test.go [hunk 3]
```diff
@@ -204,15 +205,17 @@ func TestAdminServer_UpdateTree(t *testing.T) {
 			continue
 		}
 
-		if tree.CreateTimeMillisSinceEpoch > tree.UpdateTimeMillisSinceEpoch {
-			t.Errorf("%v: CreateTime > UpdateTime (%v > %v)", test.desc, tree.CreateTimeMillisSinceEpoch, tree.UpdateTimeMillisSinceEpoch)
+		created, _ := ptypes.Timestamp(tree.CreateTime)
```
Comment (AlCutter) [inline integration/admin/admin_integration_test.go]: Should `err` be checked here?

Comment (daviddrysdale) [inline integration/admin/admin_integration_test.go]: Done, and fixed up rebase merge conflicts for the fifth time.

**Pull request merged**